### ISSUE TITLE
Klarna shipping methods

### ISSUE THREAD

**Issue by wilmer05** [elleracompany/commerce-klarna-checkout#47]
Hello I was trying to see on the code how to add shipping methods to the iframe of klarna (in the documentation from klarna is stated that one is able to show that on the iframe), but I don't see any option on the plugin to add them, is there any possibility for this?

Best regards

**Comment by jellingsen** [comment]
I don't fully understand. The chosen shipping method is shown as a order line in the Checkout and HPP gateways.

Can you link me the documentation you are referring to?

**Comment by wilmer05** [comment]
Exactly, but the shipping price strictly depend on the address where the things are going to be sent, However I'm not able to know this address (before the user introduces the address on the iframe from klarna), and when the user already introduced that, then it's already too late since I don't get any event to update the order for example. 

https://developers.klarna.com/documentation/merchant-shipping/api-integration/#3-shipping-options

**Comment by jellingsen** [comment]
Gotcha.

Taking a quick look at it, passing the shipping options alone will not solve your issue. We need to also create a callback endpoint for Klarna to call each time the address is updated in the Klarna-widget.

I will have a look at this next week, and I'll update you with an ETA then.

https://developers.klarna.com/api/#checkout-api-create-a-new-order

**Comment by wilmer05** [comment]
Yes, that Is exactly what I need. As stated here https://developers.klarna.com/api/#checkout-api-callbacks-shipping-option-update you just need to provide that option in the merchant urls ( I mean here https://github.com/elleracompany/commerce-klarna-checkout/blob/master/src/models/forms/CheckoutForm.php ) and add the option on the settings so that I can put whatever callback endpoint I want (in that way is more flexible and I can define my own endpoint)

**Comment by wilmer05** [comment]
By the way, the way it works is in the admin of Klarna one has to add a shipping merchant option... then Somehow (I still don't know since I'm waiting klarna to check my id)  add the name of the endpoint that the iframe calls when the address is updated, there we obtain the address and we can provide the shipping methods, then the iframe updates everything. At the end, klarna calls with the correct shipping method (selected by the user) the website and we can update that in craft

**Comment by wilmer05** [comment]
Hello, Just a small update, after talking to klarna team I figured out one can have the shipping methods from the configs and the iframe. However as in the previous comments I made we need an endpoint that they call when the user update the shipping address and shipping method (see two comments above)

In order to do the shipping methods options on the iframe one just have to follow this steps https://webshipper.zendesk.com/hc/en-us/articles/360010076857-Get-started-Klarna then the shipping methods appear :). I still have the problem that the orders doesn't get update when the user udpates the address or shipping option which is what this issue is about, please let me know if I can help somehow in order to do this

Best regards

**Comment by jellingsen** [comment]
Okey, let me see if i understand you correctly.

There are two ways to handle variable shipping costs in klarna:
1. You provide a shipping endpoint on the site that klarna can call when the shipping address is updated
2. You add shipping providers inside the klarna portal and create an endpoint for klarna to hit with the shipping method selected in the checkout process

You would prefer option 2?

**Comment by wilmer05** [comment]
Let me update everything a bit. At the beginning I thought we could provided through and endpoint, but the way its done is the following. 

I have to set in webshipper an integration with klarna ( and viceversa ) this can be done just wit hthe admin interfaces of both sites.

Now what you have to provide at the time of generating the form is the merchant_urls for all the gateways forms you have, for example here https://github.com/elleracompany/commerce-klarna-checkout/blob/master/src/models/forms/HostedForm.php

in that list you have to provide the option shipping_option_update as stated here https://developers.klarna.com/api/#checkout-api-callbacks-shipping-option-update so klarna will call that endpoint to tell me that the order was updated and then we can update the order lines (the shipping option) on craft. 

So what I would need is that you update the code providing that option, and to get the value of that url from the settings, so that I can put any callback url that I want and then receive their requests

**Comment by wilmer05** [comment]
So to sum up, is just adding that url to the merchant_urls, and having a variable in the environment or in the settings which would be that the new url (callback) is. With that I can continue well. Please let me know when you can fix it, sorry for the pressure but we need this as soon as possible due to our client

Best regards

**Comment by jellingsen** [comment]
I understand. As this is a plugin currently in production mode with several merchants using it, I wont just add on a extra setting for this purpose. We will add this feature, but it must be done correctly catering to all needs and configuration options that Klarna provides.

In the meantime, I have a temporary fix for you so that you can deliver for your customer:

1. Create a folder in your craft root, ex. `overrides`, and copy over the three files in `vendor/elleracompany/commerce-klarna-checkout/src/models/forms`.
2. Update your composer.json with the following changes:
``` 
  "autoload": {
    "psr-4": {
      "ellera\\commerce\\klarna\\models\\forms\\": "overrides"
    },
    "exclude-from-classmap": [
      "vendor/elleracompany/commerce-klarna-checkout/src/models/forms/BasePaymentForm.php",
      "vendor/elleracompany/commerce-klarna-checkout/src/models/forms/HostedForm.php",
      "vendor/elleracompany/commerce-klarna-checkout/src/models/forms/CheckoutForm.php"
    ]
  },
```
3. run `composer dump-autoload`
4. Update the required gateway form, ex. CheckoutForm, in your overrides folder and add the merchant urls you need in `populate()`-function near the top.
5. You should now be able to pass a new merchant url to Klarna and get your shipping working.
**Note: you need to be careful when updating this plugin in the feature, I would recommend version restricting it to the current version and update manually when a new version is published.**

**Comment by wilmer05** [comment]
Hello thanks for the solution.

but is it not possible just to make it an optional parameter? and if an env variable is set then you add the option to the merchant_urls, that way you would be breaking the others clients code

Best regards
